ext/geos_c_impl/factory.c

```diff
@@ -863,9 +864,11 @@ rgeo_convert_to_detached_geos_geometry(VALUE obj,
                               ID2SYM(rb_intern("force_new")),
                               ID2SYM(rb_intern("keep_subtype")));
   if (*state || NIL_P(object)) {
-    return NULL;
+    rb_raise(rb_eRGeoInvalidGeometry,
+             "Unable to cast the geometry to the GEOS Factory");
```

Comment:
For (1.), if we are scared of using `rb_set_errinfo` (because `rb_protect` actually seems to do much more), we could:

```c
  if (NIL_P(object))
    rb_protect(rb_exc_raise, rb_exc_new_str("Unable to cast the geometry to the GEOS Factory"), state);

  if (*state)
    return NULL;
```

Reply:
I like this solution. We'd have to do a `NULL` check but at least we can raise right then and halt execution instead of letting it propagate.

Reply:
I'll look at this soon and see if any other issues come up with it, but this seems like a fairly straightforward fix. Do you suggest doing this for both of the `geos_convert_*` methods or just the `detached` version?

Reply:
I guess both of them get used in a similar way that we see as an issue here so I can do both.

Reply:
@BuonOmo I'm going through this a little more now and I'm not sure we can meet the guarantee of the `rgeo_convert_*` methods not raising since we use `Check_TypedStruct` to validate the data types which may raise.

I can modify the `detached` version so that if it's `NULL` or `state` is returned, the user should handle the error, and the `state` variable doesn't just disappear, but I don't think it makes sense to pull the `Check_TypedStruct` calls out of these functions. What do you think?

Reply:
https://github.com/rgeo/rgeo/commit/6f241060f45d0477e23492794c48e789d0dd3190 This is the commit where we a added that check btw.

Reply:
@BuonOmo looked at this some more. It seems like I can replace these `Check_TypedStruct` calls with the `rgeo_is_geos_object` or `rgeo_check_geos_object` functions we defined and wrap it in `rb_protect`.

Reply:
@keithdoggett it seems like there was a lot more to dig than what I expected. Maybe raising is the simpler option?

If you feel like you have the problem under control however, I fully agree with your choices.

My last concern is about performances, they may be a bit degraded when using protect. But tbh I don't know, and I'm not sure for now how to handle that.

Reply:
Yeah I can look into `rb_protect` and see if performance is going to be really degraded. I think `rb_protect` is intended for calling ruby methods from C code, but the interface is much nicer since you don't have to define a callback function with specific args like you do with `rb_rescue`.

I would say from a "coding flow" standpoint it feels good right now and the 2 `rgeo_convert_*` functions have a clear error handling pattern through `state`, but I can investigate `rb_protect`.